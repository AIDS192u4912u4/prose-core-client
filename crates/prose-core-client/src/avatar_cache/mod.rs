pub use avatar_cache::{AvatarCache, IMAGE_OUTPUT_MIME_TYPE, MAX_IMAGE_DIMENSIONS};
pub use noop_avatar_cache::NoopAvatarCache;

mod avatar_cache;
mod noop_avatar_cache;

#[cfg(feature = "native-app")]
pub mod fs_avatar_cache;
