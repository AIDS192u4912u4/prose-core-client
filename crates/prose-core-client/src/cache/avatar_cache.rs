use std::path::PathBuf;

use anyhow::Result;
use jid::BareJid;

use prose_xmpp::stanza::avatar;

pub const MAX_IMAGE_DIMENSIONS: (u32, u32) = (600, 600);
pub const IMAGE_OUTPUT_MIME_TYPE: &str = "image/jpeg";

pub trait AvatarCache: Send + Sync {
    #[cfg(feature = "native-app")]
    fn cache_avatar_image(
        &self,
        jid: &BareJid,
        image: image::DynamicImage,
        metadata: &crate::types::AvatarMetadata,
    ) -> Result<PathBuf>;

    fn cached_avatar_image_url(
        &self,
        jid: &BareJid,
        image_checksum: &avatar::ImageId,
    ) -> Option<PathBuf>;

    fn delete_all_cached_images(&self) -> Result<()>;
}
