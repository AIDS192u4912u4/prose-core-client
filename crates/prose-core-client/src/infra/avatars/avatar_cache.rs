// prose-core-client/prose-core-client
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use anyhow::Error;
use async_trait::async_trait;

use prose_wasm_utils::{SendUnlessWasm, SyncUnlessWasm};
use prose_xmpp::mods::AvatarData;

use crate::domain::shared::models::{AccountId, AvatarId, UserId};
use crate::domain::user_info::models::{AvatarInfo, PlatformImage};

pub const MAX_IMAGE_DIMENSIONS: (u32, u32) = (400, 400);

#[cfg_attr(target_arch = "wasm32", async_trait(? Send))]
#[async_trait]
pub trait AvatarCache: SendUnlessWasm + SyncUnlessWasm {
    async fn cache_avatar_image(
        &self,
        account: &AccountId,
        user_id: &UserId,
        image: &AvatarData,
        metadata: &AvatarInfo,
    ) -> Result<(), Error>;

    async fn has_cached_avatar_image(
        &self,
        account: &AccountId,
        user_id: &UserId,
        image_checksum: &AvatarId,
    ) -> Result<bool, Error>;

    async fn cached_avatar_image(
        &self,
        account: &AccountId,
        user_id: &UserId,
        image_checksum: &AvatarId,
    ) -> Result<Option<PlatformImage>, Error>;

    async fn delete_all_cached_images(&self, account: &AccountId) -> Result<(), Error>;
}
