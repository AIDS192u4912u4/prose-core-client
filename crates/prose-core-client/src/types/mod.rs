// pub use message::Message;
pub use account_settings::AccountSettings;
pub use availability::Availability;
pub use avatar_metadata::AvatarMetadata;
pub use capabilities::{Capabilities, Feature};
pub use contact::Contact;
pub use message::{Emoji, Message, MessageId, Reaction, StanzaId};
pub use message_like::MessageLike;
pub use page::Page;
pub use presence::Presence;
pub use user_activity::UserActivity;
pub use user_metadata::UserMetadata;
pub use user_profile::{Address, Url, UserProfile};

mod account_settings;
mod availability;
mod avatar_metadata;
mod capabilities;
mod contact;
mod error;
mod message;
pub mod message_like;
mod page;
pub mod presence;
pub mod roster;
mod user_activity;
pub mod user_metadata;
mod user_profile;
