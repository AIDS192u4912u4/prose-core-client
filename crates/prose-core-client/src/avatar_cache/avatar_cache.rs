// prose-core-client/prose-core-client
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use async_trait::async_trait;
#[cfg(target_arch = "wasm32")]
use auto_impl::auto_impl;
use jid::BareJid;

use prose_xmpp::mods::AvatarData;
use prose_xmpp::stanza::avatar;
use prose_xmpp::{SendUnlessWasm, SyncUnlessWasm};

use crate::types::AvatarMetadata;

pub const MAX_IMAGE_DIMENSIONS: (u32, u32) = (400, 400);

#[cfg_attr(target_arch = "wasm32", async_trait(? Send), auto_impl(Rc))]
#[async_trait]
pub trait AvatarCache: SendUnlessWasm + SyncUnlessWasm {
    type Image;
    type Error: std::error::Error + Send + Sync;

    async fn cache_avatar_image(
        &self,
        jid: &BareJid,
        image: &AvatarData,
        metadata: &AvatarMetadata,
    ) -> Result<(), Self::Error>;

    async fn has_cached_avatar_image(
        &self,
        jid: &BareJid,
        image_checksum: &avatar::ImageId,
    ) -> Result<bool, Self::Error>;

    async fn cached_avatar_image(
        &self,
        jid: &BareJid,
        image_checksum: &avatar::ImageId,
    ) -> Result<Option<Self::Image>, Self::Error>;

    async fn delete_all_cached_images(&self) -> Result<(), Self::Error>;
}
