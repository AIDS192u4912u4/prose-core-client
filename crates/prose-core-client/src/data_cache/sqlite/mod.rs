// prose-core-client
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use std::str::FromStr;

use rusqlite::types::{FromSql, FromSqlError, FromSqlResult, ValueRef};

pub use cache::{Connection, SQLiteCache, SQLiteCacheError};

mod cache;
mod contacts_cache;
mod message_cache;

pub(self) fn repeat_vars(count: usize) -> String {
    let mut s = "?,".repeat(count);
    // Remove trailing comma
    s.pop();
    s
}

pub(self) struct FromStrSql<T>(pub T);

impl<T: FromStr> FromSql for FromStrSql<T> {
    fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self> {
        T::from_str(value.as_str()?)
            .map_err(|_| FromSqlError::InvalidType)
            .map(FromStrSql)
    }
}
