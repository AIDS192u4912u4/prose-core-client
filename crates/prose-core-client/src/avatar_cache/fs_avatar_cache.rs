use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Error, Result};
use image::{DynamicImage, ImageOutputFormat};
use jid::BareJid;

use prose_xmpp::stanza::avatar;

use crate::avatar_cache::AvatarCache;
use crate::types::AvatarMetadata;

pub const IMAGE_OUTPUT_FORMAT: ImageOutputFormat = ImageOutputFormat::Jpeg(80);

pub struct FsAvatarCache {
    path: PathBuf,
}

impl FsAvatarCache {
    pub fn new(path: &Path) -> Result<Self> {
        fs::create_dir_all(&path)?;

        Ok(FsAvatarCache {
            path: path.to_path_buf(),
        })
    }
}

impl AvatarCache for FsAvatarCache {
    fn cache_avatar_image(
        &self,
        jid: &BareJid,
        image: DynamicImage,
        metadata: &AvatarMetadata,
    ) -> Result<PathBuf> {
        let output_path = self.path.join(self.filename_for(jid, &metadata.checksum));
        let mut output_file = std::fs::File::create(&output_path)?;

        // Sometimes we encounter e.g. rgb16 pngs and image-rs complains that the JPEG encoder
        // cannot save these, so we convert the image to rgb8.
        image
            .into_rgb8()
            .write_to(&mut output_file, IMAGE_OUTPUT_FORMAT)?;
        Ok(output_path)
    }

    fn cached_avatar_image_url(
        &self,
        jid: &BareJid,
        image_checksum: &avatar::ImageId,
    ) -> Option<PathBuf> {
        let path = self.filename_for(jid, image_checksum);
        if path.exists() {
            return Some(path);
        }
        return None;
    }

    fn delete_all_cached_images(&self) -> Result<()> {
        for entry in fs::read_dir(&self.path)? {
            let entry = match entry {
                Ok(entry) => entry,
                Err(err) => return Err(Error::new(err)),
            };
            let metadata = entry.metadata()?;
            if metadata.is_file()
                && entry.path().extension().and_then(|ext| ext.to_str()) == Some("jpg")
            {
                fs::remove_file(entry.path())?
            }
        }
        Ok(())
    }
}

impl FsAvatarCache {
    fn filename_for(&self, jid: &BareJid, image_checksum: &avatar::ImageId) -> PathBuf {
        self.path.join(format!(
            "{}-{}.jpg",
            jid.to_string(),
            image_checksum.as_ref()
        ))
    }
}
