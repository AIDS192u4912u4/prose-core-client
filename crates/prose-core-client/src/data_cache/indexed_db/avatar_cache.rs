// prose-core-client
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use async_trait::async_trait;
use indexed_db_futures::prelude::*;
use jid::BareJid;
use prose_xmpp::mods::AvatarData;
use wasm_bindgen::JsValue;

use crate::avatar_cache::AvatarCache;
use prose_xmpp::stanza::avatar::ImageId;

use crate::data_cache::indexed_db::cache::{keys, IndexedDBDataCacheError};
use crate::data_cache::indexed_db::idb_database_ext::{IdbDatabaseExt, IdbObjectStoreExtGet};
use crate::data_cache::indexed_db::IndexedDBDataCache;
use crate::types::AvatarMetadata;

use super::cache::Result;

#[async_trait(? Send)]
impl AvatarCache for IndexedDBDataCache {
    type Image = String;
    type Error = IndexedDBDataCacheError;

    async fn cache_avatar_image(
        &self,
        _jid: &BareJid,
        image: &AvatarData,
        metadata: &AvatarMetadata,
    ) -> Result<()> {
        self.db
            .set_value(
                keys::AVATAR_STORE,
                &metadata.checksum,
                image.base64().as_ref(),
            )
            .await?;
        Ok(())
    }

    async fn has_cached_avatar_image(
        &self,
        _jid: &BareJid,
        image_checksum: &ImageId,
    ) -> Result<bool> {
        let tx = self
            .db
            .transaction_on_one_with_mode(keys::AVATAR_STORE, IdbTransactionMode::Readonly)?;
        let avatar_store = tx.object_store(keys::AVATAR_STORE)?;
        let key_exists = avatar_store
            .get_key(&JsValue::from_str(image_checksum.as_ref()))?
            .await?
            .is_some();
        Ok(key_exists)
    }

    async fn cached_avatar_image(
        &self,
        jid: &BareJid,
        image_checksum: &ImageId,
    ) -> Result<Option<Self::Image>> {
        let tx = self.db.transaction_on_multi_with_mode(
            &[keys::AVATAR_METADATA_STORE, keys::AVATAR_STORE],
            IdbTransactionMode::Readonly,
        )?;

        let avatar_metadata_store = tx.object_store(keys::AVATAR_METADATA_STORE)?;
        let avatar_store = tx.object_store(keys::AVATAR_STORE)?;

        let avatar_metadata = avatar_metadata_store
            .get_value::<AvatarMetadata>(jid.to_string())
            .await?;
        let base64_data = avatar_store.get_value::<String>(image_checksum).await?;

        let (Some(avatar_metadata), Some(base64_data)) = (avatar_metadata, base64_data) else {
            return Ok(None);
        };

        let data_url = format!("data:{};base64,{}", avatar_metadata.mime_type, base64_data);

        Ok(Some(data_url))
    }

    async fn delete_all_cached_images(&self) -> Result<()> {
        self.db.clear_stores(&[keys::AVATAR_STORE]).await?;
        Ok(())
    }
}
