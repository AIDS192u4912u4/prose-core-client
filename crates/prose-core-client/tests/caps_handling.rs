// prose-core-client/prose-core-client
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use anyhow::Result;
use insta::assert_snapshot;
use jid::{BareJid, FullJid};
use prose_core_client::avatar_cache::NoopAvatarCache;
use prose_core_client::data_cache::sqlite::SQLiteCache;
use prose_core_client::test::{ClientTestAdditions, ConnectedClient};
use prose_core_client::types::{Availability, SoftwareVersion};
use prose_core_client::{Client, ClientBuilder};
use prose_xmpp::test::{BareJidTestAdditions, IncrementingIDProvider};
use prose_xmpp::{test, SystemTimeProvider};
use std::str::FromStr;
use std::sync::Arc;

// Snapshots will need to be updated if/when caps features changeâ€¦

#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
async fn test_start_sequence() -> Result<()> {
    let connection = Arc::new(test::Connection::default());
    let data_cache = Arc::new(SQLiteCache::in_memory_cache());

    let client = ClientBuilder::new()
        .set_connector_provider(test::Connector::provider(connection.clone()))
        .set_data_cache(data_cache.clone())
        .set_avatar_cache(NoopAvatarCache::default())
        .set_id_provider(IncrementingIDProvider::new())
        .set_time_provider(SystemTimeProvider::default())
        .set_software_version(SoftwareVersion {
            name: "prose-test-client".to_string(),
            version: "1.2.3".to_string(),
            os: Some("unknown os".to_string()),
        })
        .build();

    client
        .connect(
            &FullJid::from_str(&format!("{}/test", BareJid::ours()))?,
            "",
            Availability::Available,
        )
        .await?;

    assert_snapshot!(connection.sent_stanza_strings().join("\n\n"));

    Ok(())
}

#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
async fn test_sends_caps_when_changing_availability() -> Result<()> {
    let ConnectedClient {
        client, connection, ..
    } = Client::connected_client().await?;

    client.set_availability(Availability::DoNotDisturb).await?;

    let sent_stanzas = connection.sent_stanza_strings();
    assert_eq!(sent_stanzas.len(), 1);
    assert_snapshot!(sent_stanzas[0]);

    Ok(())
}
