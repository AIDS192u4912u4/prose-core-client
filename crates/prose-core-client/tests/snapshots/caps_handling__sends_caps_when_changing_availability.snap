---
source: crates/prose-core-client/tests/caps_handling.rs
assertion_line: 63
expression: "sent_stanzas[0]"
---
<presence xmlns='jabber:client'><show>dnd</show><c xmlns='http://jabber.org/protocol/caps' hash="sha-1" node="sha-1" ver="aHR0cHM6Ly9wcm9zZS5vcmc="/></presence>
